crates/transaction-pool/src/maintain.rs

```diff
@@ -64,17 +95,100 @@ where
     // keeps track of the state of the pool wrt to blocks
     let mut maintained_state = MaintainedPoolState::InSync;
 
+    // the future that reloads accounts from state
+    let mut reload_accounts_fut = Fuse::terminated();
+
+    // The update loop that waits for new blocks and reorgs and performs pool updated
     // Listen for new chain events and derive the update action for the pool
     loop {
+        trace!(target = "txpool", state=?maintained_state, "awaiting new block or reorg");
+
         metrics.set_dirty_accounts_len(dirty_addresses.len());
+        let pool_info = pool.block_info();
 
-        let Some(event) = events.next().await else { break };
+        // after performing a pool update after a new block we have some time to properly update
+        // dirty accounts and correct if the pool drifted from current state, for example after
+        // restart or a pipeline run
+        if maintained_state.is_drifted() {
+            // assuming all senders are dirty
+            dirty_addresses = pool.unique_senders();
+            maintained_state = MaintainedPoolState::InSync;
+        }
 
-        let pool_info = pool.block_info();
+        // if we have accounts that are out of sync with the pool, we reload them in chunks
+        if !dirty_addresses.is_empty() && reload_accounts_fut.is_terminated() {
+            let (tx, rx) = oneshot::channel();
+            let c = client.clone();
+            let at = pool_info.last_seen_block_hash;
+            let fut = if dirty_addresses.len() > max_reload_accounts {
+                // need to chunk accounts to reload
+                let accs_to_reload =
+                    dirty_addresses.iter().copied().take(max_reload_accounts).collect::<Vec<_>>();
+                for acc in &accs_to_reload {
+                    // make sure we remove them from the dirty set
+                    dirty_addresses.remove(acc);
+                }
+                async move {
+                    let res = load_accounts(c, at, accs_to_reload.into_iter());
+                    let _ = tx.send(res);
+                }
+                .boxed()
+            } else {
+                // can fetch all dirty accounts at once
+                let accs_to_reload = std::mem::take(&mut dirty_addresses);
+                async move {
+                    let res = load_accounts(c, at, accs_to_reload.into_iter());
+                    let _ = tx.send(res);
+                }
+                .boxed()
+            };
+            reload_accounts_fut = rx.fuse();
+            task_spawner.spawn_blocking(fut);
```

Comment:
makes sense - `load_accounts` has `basic_account` calls in it which are blocking. this is still too easy to miss, how can we make the blocking calls inside of a function more transparent to consumers?